### PR TITLE
civicrm_cache - Replace wonky index

### PR DESCRIPTION
**Pull request by totten**
Overview
----------------------------------------

The `civicrm_cache` table stores groups of key-value pairs. It includes a uniqueness constraint. While _some_ `UNIQUE` constraint is appropriate here, the _specific_ constraint doesn't make sense.

Before
----------------------------------------

* `civicrm_cache` is unique based on `group_name, path, created_date`.

After
----------------------------------------

* `civicrm_cache` is unique based on `group_name, path`.

Comments
----------------------------------------

Uniqueness constraint for `group_name,path` is consistent with the shape of the PHP API and the behavior of the other cache implementations. The question is: why was `created_date` in there? I traced it to:

* https://github.com/civicrm/civicrm-svn/commit/cfe0fd611fb65391caf44e4c45e9490c9b2f75d5
* https://issues.civicrm.org/jira/browse/CRM-10335

It appears to me that it originally did use `group_name,path`. The `,created_date` was added in a response to a request to improve performance. This conflates two purposes - the *logic* of a uniqueness constraint and the *performance* of a date index. But I don't see how this status-quo serves either purpose well:

* By adding `created_date`, the uniqueness constraint no longer serves the logical purpose of preventing duplicates. Instead, to prevent duplicates, `SqlGroup::set()` requires extra locking and multiple queries.
* BTREE's aren't magic. If you look at [the queries used circa 4.x](https://github.com/civicrm/civicrm-core/blob/4.7.30/CRM/Core/BAO/Cache.php#L328-L331), they filtered on `group_name` and `created_date`. But if the BTREE covers `group_name => path => created_date`, then the intervening `path` would seem to limit its value. (*However, some of the attachments from CRM-10335 aren't available, so I can't fully see what they were reacting to.*)
* In any event, the cleanup process changed. It no longer uses `group_name` or `created_date` at all. Since 5.4, it has [used `expired_date` and only `expired_date`](https://github.com/civicrm/civicrm-core/blob/5.4/CRM/Core/BAO/Cache.php#L369).

### CODE REVIEW

File: CRM/Upgrade/Incremental/php/FiveSixtyOne.php [hunk 1]
```diff
@@ -46,6 +46,38 @@ public function setPreUpgradeMessage(&$preUpgradeMessage, $rev, $currentVer = NU
...
+   * duplicates are somehow in there, then we should cleanly remove them rather than let the upgrade fail.
+   *
+   * @param \CRM_Queue_TaskContext $ctx
+   * @return bool
+   * @throws \Civi\Core\Exception\DBQueryException
+   */
+  public static function dedupeCache($ctx): bool {
+    $duplicates = CRM_Core_DAO::executeQuery('
+      SELECT c.id FROM civicrm_cache c
+      LEFT JOIN (SELECT group_name, path, max(created_date) newest FROM civicrm_cache GROUP BY group_name, path) recent
```
Comment (eileenmcnaughton) [inline CRM/Upgrade/Incremental/php/FiveSixtyOne.php:69]: Hmm - you didn't think about just truncating the cache?

**Comment by eileenmcnaughton** [comment]
The cleanup query is `  $sql = "DELETE FROM civicrm_cache WHERE expired_date < %1";`

So - the right index for performance is an index on `expired_date` - it might be worth adding at the same time?

I generally agree the unique index as it is makes no sense

File: CRM/Upgrade/Incremental/php/FiveSixtyOne.php (hunk 1 above)
Comment (totten) [reply CRM/Upgrade/Incremental/php/FiveSixtyOne.php:69]: Yeah, I briefly considered that there are several formulations that would probably do the job but didn't linger on it. I'll grant this formulation is more code-heavy than others, but it's one-off, and - behaviorally speaking - it's the lightest formulation (i.e. less likely to have some unforeseen interaction). Overall, it's probably 6/half-dozen.

FWIW, I did do some `r-run` with redundant records to ensure that it got the right ones.

**Comment by totten** [comment]
> So - the right index for performance is an index on expired_date - it might be worth adding at the same time?

Yeah, I'm happy to throw that in. I had the same thought. I know that y'all are sometimes skeptical about adding indexes, but if you're good with it, then I am too. It is a fairly simple/single-column thing.

**Comment by eileenmcnaughton** [comment]
SO risks on adding indexes are

1) adding an index to a large table can make upgrading slow (not relevant)
2) they can be a bit flakey to add - sometimes it fails based on stuff we are not that clear about
3) sometimes they don't really perform better. I don't think that applies here but for example an index on contribution_status_id can actually hurt performance except when specifically searching for a less common status

**Comment by totten** [comment]
> sometimes they don't really perform better ... contribution_status_id ...

Ah, right, that's not generally a very distinctive key (*you get 4-5 massive buckets*). By contrast, indexing on `expired_date`  should give a meaningful split between "too old/expired" and "newish/unexpired".

**Comment by totten** [comment]
Added the `expired_date`. FWIW, when I run `cv upgrade:db` (5.60=>master) or `civibuild reinstall`, I get similar schema:

```sql
  PRIMARY KEY (`id`),
  UNIQUE KEY `UI_group_name_path` (`group_name`,`path`),
  KEY `index_expired_date` (`expired_date`),
  KEY `FK_civicrm_cache_component_id` (`component_id`),
  CONSTRAINT `FK_civicrm_cache_component_id` FOREIGN KEY (`component_id`) REFERENCES `civicrm_component` (`id`)
```

(*key names/types/columns were matched-up; line-order may vary trivially*)

**Comment by eileenmcnaughton** [comment]
OK - I'm happy with the testing you have done & agree this was good sleuthing / resolution to the WTF moment you obviously experienced

**Pull request merged**